### DISCUSSION TITLE
Loops in Trigger Functions

### DISCUSSION THREAD

**Discussion by ghost** [supabase/supabase#14020] [Questions] [answered]
Disclaimer:  I have about 48 hours worth of experience with PostgreSQL and Supabase.

I have 3 tables in my database:
- Default Machine Order Steps (currently 3 rows)
- Machine Orders
- Project Steps

Goal is that every time a new Machine Order is created, 3 new Project Steps are created based on all the rows in the Default Machine Order Steps.  Each Project Step has a foreign key to the triggering Machine order.  Here's the code:

```plpgsql
declare
  step record;
begin
  for step in (
    select "Title" from "Default Machine Order Steps"
  ) loop 
    insert into "Project Steps"("Title","Status", "Machine Order", "Machine Order Name")
    values(step."Title",'Not Started', new.id, new."Title");
    return new;

  end loop;
end;
``` 

Return type is trigger, language is plpgsql.

As it stand the code triggers and runs, but only 1 row is inserted.  I assume this is due to `return new;`.  When I change this to `return next;`  like what was originally used in testing on the SQL editor, I get the following error:

`Failed to create function: failed to update pg.functions with the given ID: cannot use RETURN NEXT in a non-SETOF function`

Thanks in advance for any insights!

**Comment by GaryAustin1** [comment, marked as answer]
You can loop and do operations on other tables in a trigger function, but there should only be one return new at the very end of the trigger.

**Reply by ghost** [reply]
Awesome.  That was it--thank you.

Correct code:

```plpgsql
declare
  step record;
begin
  for step in (
    select "Title" from "Default Machine Order Steps"
  ) loop 
    insert into "Project Steps"("Title","Status", "Machine Order", "Machine Order Name")
    values(step."Title",'Not Started', new.id, new."Title");
    

  end loop;
  return new;
end;
```
